### ISSUE TITLE
Neural Net doesn't learn. Vanishing gradients?

### ISSUE THREAD

**Issue by psilva07** [dmlc/MXNet.jl#241]
I have a dataset with 181 features and about 700000 samples and I've been trying to perform regression on it. The dataset is properly standardized, shuffled and doesn't contain outliers. I have explored it in depth with random forests and I got to a decent place but I believe I can get better results with Neural Nets. I used to work with Mocha.jl but they say we should be moving to MXNet so here I am. 

For whatever reason, my MX Nets don't seem to learn. I've tried all the different initializers, activation functions, optimizers, architectures with 1 and 2 hidden layers, from 100 to 2000 hidden units, and the results are still the same. After setting everything up and training dozens of different models I'm at a loss here. My best guess is that there is some vanishing gradient problem occurring but I'm not sure how to deal with it, and why it only occurs with MXNet. The net starts with an RMSE of about .45 and "finishes" training after (500,1000,5000) epochs, with and RMSE of .43 or .44. Training similar models on Mocha.jl, gives me a final RMSE of .15, which is close to what I had with the random forests (.13).

I've been checking the range of the outputs and although the real output should be between 0 and 1, the outputs of the net are between .46 and .75. Even when I subsample from the dataset to make it have the same amount of objects in n buckets between 0 and 1. For example, I can have 5 buckets (objects where the output is between 0.0-0.2; 0.2-0.4; 0.4-0.6; 0.6-0.8; 0.8-1.0) with 20000 objects each, and the same problem persists. 

My code looks very much like the regression example:

```julia
TrainInput = fullDataSTD[1:615000,:]'
TrainOutput = Ysum[1:615000]
ValidationInput = fullDataSTD[615001:end,:]'
ValidationOutput = Ysum[615001:end];

# how to set up data providers using data in memory
batchsize = 1000 # can adjust this later, but must be defined now for next line
trainprovider = mx.ArrayDataProvider(:data => TrainInput, batch_size=batchsize, shuffle=true, :label => TrainOutput)
evalprovider = mx.ArrayDataProvider(:data => ValidationInput, batch_size=batchsize, shuffle=true, :label => ValidationOutput)

data = mx.Variable(:data)
label = mx.Variable(:label)
net = @mx.chain     mx.Variable(:data) =>
                    mx.FullyConnected(num_hidden=500) =>
                    mx.Activation(act_type=:softrelu) =>
                    #mx.FullyConnected(num_hidden=100) =>
                    #mx.Activation(act_type=:sigmoid) =>
                    mx.FullyConnected(num_hidden=1) =>        
                    mx.LinearRegressionOutput(label)

cost = mx.LogisticRegressionOutput(mx.SymbolicNode, data = net, label=label)
model = mx.FeedForward(cost, context=mx.cpu())
optimizer = mx.ADAM()

# train, reporting loss for training and evaluation sets
# initial training with small batch size, to get to a good neighborhood
batchsize = 200
mx.fit(model, optimizer, initializer=mx.NormalInitializer(0.0,0.1), eval_metric=mx.MSE(), trainprovider, eval_data=evalprovider, n_epoch = 20)
#mx.fit(model, optimizer, trainprovider, n_epoch=20, eval_data=evalprovider, eval_metric=mx.Accuracy())
# more training with the full sample
batchsize = samplesize
mx.fit(model, optimizer, eval_metric=mx.MSE(), trainprovider, eval_data=evalprovider, n_epoch = 5000)

# obtain predictions
plotprovider = mx.ArrayDataProvider(:data => ValidationInput, :label => ValidationOutput)
fit = mx.predict(model, plotprovider)
println("RMSE: $(root_mean_squared_error(ValidationOutput', fit))")
```

Any help is welcome.

**Comment by vchuravy** [comment]
This seems odd to me:
```
net = ... mx.LinearRegressionOutput(label)
cost = mx.LogisticRegressionOutput(mx.SymbolicNode, data = net, label=label)
```

Why do you want to first take a `LinearRegression` with respect to `label` and then a `LogisticRegression` with respect to label?

What happens if you say: `model = mx.FeedForward(net, context=mx.cpu())`

**Comment by psilva07** [comment]
That's it! Such a simple mistake. I was doing classification before, on different labels, and I mixed up the code afterward, or so it seems. I'm happy that it was such a simple error. I'm also sad for the same reason. Thank you very much, @vchuravy !
